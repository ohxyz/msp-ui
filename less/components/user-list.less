@import "./message-box";

.user-list {

    margin: 0 auto;
    width: 750px;
    padding-left: 15px;
    padding-right: 15px;
    padding-top: 24px;

    &__container {

        background-color: #f1f1f1;
        min-height: 600px;
    }

    &__count {

        margin-bottom: 24px;

        .material-icons {

            color: #ffb432;
            font-size: 24px;
        }

        &__literal {

            font-size: 14px;
            line-height: 24px;
            color: #505050;
            display: inline-block;
            position: relative;
            top: -6px;
            margin-left: 8px;
        }
    }
}

.user-strip {

    margin-bottom: 24px;

    &__main {

        position: relative;
        background-color: white;
        padding: 24px 64px 20px 16px;
    }

    &__delete-icon, 
    &__expand-icon {

        position: absolute;
        display: flex;
        justify-content: center;
        align-items: center;
        cursor: pointer;
        width: 48px;
        line-height: 40px;
        right: 0;
        user-select: none;
    }

    &__delete-icon {

        bottom: 50%;
        top: 0;
        border-left: 1px solid #e3e3e3;
        border-bottom: 1px solid #e3e3e3;

        &--inactive {
            
            color: #ababab;
        }

        &--active {

            color: #ff373c;
        }
    }

    &__expand-icon {

        top: 50%;
        bottom: 0;
        border-left: 1px solid #e3e3e3;
    }

    &__full-name {

        display: inline-block;
        font-size: 16px;
        line-height: 24px;
        margin-right: 16px;
        font-weight: bold;
        color: #232323;
    }

    &__top-org {

        display: inline-block;
        font-size: 12px;
        line-height: 20px;
        padding: 2px 4px;
        color: white;
        background-color: #3f75c6;
    }

    &__admin{

        display: inline-block;
        font-size: 12px;
        line-height: 20px;
        margin-left: 8px;
        padding: 2px 4px;
        color: #505050;
        background-color: #f1f1f1;
    }

    &__top-bar {

        margin-bottom: 16px;
    }

    &__org {

        display: inline-block;
        font-size: 14px;
        line-height: 24px;
        color: #505050;
        margin-right: 16px;
    }

    &__more {

        display: inline-block;
        font-size: 14px;
        color: #505050;
        line-height: 24px;
        cursor: pointer;
        text-decoration: underline;
        user-select: none;

        &:hover {

            color: #ec0000;
            text-decoration: none;
        }
    }

    &__access-levels {

        background-color: #f8f8f8;
        padding: 0 16px;
        height: 0;
        transition: all 0.3s ease;

        &--active {

            padding: 24px 16px;
            height: auto;
        }
    }

    &__delete-user {

        margin-top: 8px;
        background-color: white;
        height: 0;
        transition: all 0.3s ease;

        &--active {

            height: 96px;
            padding: 24px 16px;

        }

        &__confirmation {

            display: flex;
            justify-content: center;
            font-size: 14px;
            line-height: 24px;
            color: #505050;
            margin-bottom: 24px;
        }

        &__buttons {

            display: flex;
            justify-content: center;
            height: 48px;

            button {

                box-sizing: border-box;
                width: 80px;
                height: 48px;
                font-size: 16px;
                font-weight: 500;
                border: 1px solid #ec0000;
                outline: none;
                margin: 0 8px;
                cursor: pointer;


                &:hover {

                    background-color: #b30000;
                    border: none;
                    color: white;
                }

                &:focus {

                    position: relative;
                    top: -4px;
                    height: 56px;
                    width: 88px;
                    background-color: #b30000;
                    border: 4px solid #fbc2c2;
                    margin-right: 4px;
                    margin-left: 4px;
                    color: white;
                }
            }

            &__yes {

                background-color: #ec0000;
                color: white;

            }

            &__no {

                color: #ec0000;
            }
        }
    }
}