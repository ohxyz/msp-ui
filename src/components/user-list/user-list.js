/* User List component 
 * 
 * Markups:
 * 
 * <UserList users={ AccountProfile[] } /> 
 *
 */

const React = require( 'react' );
const util = require( '../core/util.js' );
const UserStrip = require( './user-strip.js' ).UserStrip;
const componentManager = require( '../core/component-manager.js' ).componentManager;

class UserList extends React.Component {

    constructor( props ) {

        super( props );

        this.state = {

            users: [],
            shouldRenderCount: true,
            onRenderCount: new Function(),
            sortByFields: []
        };

        this.id = util.setDefault( props.id, util.generateRandomString() );
        componentManager.addComponent( this.id, this );
    }

    static getDerivedStateFromProps( nextProps, prevState ) {
        
        return {

            users: util.setDefault( nextProps.users, [] ),
            shouldRenderCount: util.setDefault( nextProps.shouldRenderCount, true ),
            onRenderCount: util.setDefault( nextProps.onRenderCount, new Function() ),
            sortByFields: util.setDefault( nextProps.sortByFields, [] )
        };
    }

    sortUsersByFields( users, fieldNames ) {

        if ( Array.isArray( fieldNames) === false || fieldNames.length === 0 ) {

            return users;
        }

        function compare( a, b, fields, fieldIndex ) {

            let field = fields[ fieldIndex ];

            return a[ field ].localeCompare( b[ field ] );
        }

        return users.sort( ( a, b ) => {

            let result = 0;

            for ( let i = 0; i < fieldNames.length; i ++ ) {

                result = compare( a, b, fieldNames, i )

                if ( result !== 0 ) {

                    return result;
                }
            }

            return result;

        } );
    }

    renderCount() {

        if ( this.state.shouldRenderCount === false ) {

            return;
        }

        let count = this.state.users.length;

        return (

            <div className="user-list__count">
                <span className="user-list__count__literal">{ this.state.onRenderCount( count ) }</span>
            </div>
        );
    }

    render() {

        let users = this.sortUsersByFields( this.state.users, this.state.sortByFields );

        return (

            <div className="user-list">
            { this.renderCount() }
            {
                users.map( ( user, key ) => {

                    return <UserStrip key={ key } user={ user } />
                } )
            }
            </div>
        );
    }
}

module.exports = {

    UserList
};