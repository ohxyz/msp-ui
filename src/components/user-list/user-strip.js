const util = require( '../core/util.js' );
const React = require( 'react' );
const UserProfile = require( '../../model/user-profile.js' ).UserProfile;

function getExpandIconText( shouldExpand ) {

    return shouldExpand === true ? 'remove' : 'add';

}

class UserStrip extends React.Component {

    constructor( props ) {

        super( props );

        this.handleExpandClick = this.handleExpandClick.bind( this );

        this.state = {

            user: null,
            shouldDropDownExpanded: false,
            expandIconText: 'add'
        };
    }

    static getDerivedStateFromProps( nextProps, prevState ) {

        let shouldExpand = util.setDefault( nextProps.shouldDropDownExpanded, false );
        
        return {

            user: util.setDefault( new UserProfile( nextProps.user ), new UserProfile() ),
            shouldDropDownExpanded: shouldExpand,
            expandIconText: getExpandIconText( shouldExpand )
        };
    }

    handleExpandClick() {

        let shouldExpand = !this.state.shouldDropDownExpanded;

        this.setState( { 

            shouldDropDownExpanded: shouldExpand,
            expandIconText: getExpandIconText( shouldExpand )

        } );
    }

    renderDropDown() {

        if ( this.state.shouldDropDownExpanded === true ) {

            return (

                <div className="user-strip__dropdown">
                    <div className="user-strip__dropdown__content">User access level</div>
                </div>
            );
        }
    }

    render() {

        let user = this.state.user;

        return (

            <div className="user-strip">
                <div className="user-strip__main">
                    <div className="user-strip__delete material-icons">close</div>
                    <div 
                        className="user-strip__expand material-icons" 
                        onClick={ this.handleExpandClick }
                    >
                        { this.state.expandIconText }
                    </div>
                    <div className="user-strip__top-bar">
                        <span className="user-strip__full-name">{ user.fullName }</span>
                        <span className="user-strip__top-group">{ user.topGroup }</span>
                    </div>
                    <div className="user-strip__middle-bar">
                        <div className="user-strip__group">{ user.group }</div>
                    </div>
                </div>
                { this.renderDropDown() }
            </div>
        );
    }
}


module.exports = {

    UserStrip,
};

