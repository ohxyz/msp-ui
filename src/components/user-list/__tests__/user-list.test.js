import React from 'react';
import { UserList } from '../user-list.js';
import renderer from 'react-test-renderer';

it( 'should render with defaults', () => {

    let user1 = { 

        fullName: 'Full Name',
        topOrgName: 'Top Orgnisation Name',
        orgName: 'Orgnisation Name'
    };

    let user2 = { 

        fullName: 'Full Name 2',
        topOrgName: 'Top Orgnisation Name 2',
        orgName: 'Orgnisation Name 2'
    };

    let users = [ user1, user2 ];
    
    const tree = renderer
        .create( <UserList users={ users }/> )
        .toJSON();

    expect( tree ).toMatchSnapshot();

} );

describe( 'UserList Instance', () => {

    let ul = new UserList( { id: 'user-list' } );
    let users = [

        { 'org': 'abcde Transport' },
        { 'org': 'abcd Education', name: 'abcd' },
        { 'org': 'abce Health', name: 'abce health people' },
        { 'name': 'Mr abcdg ' },
        { 'name': 'abcd lady', 'org': 'abce LADY GROUP' }
    ];

    test( 'sortUsersByFields method', () => { 

        let results = ul.sortUsersByFields( users, [ 'name' ] );

        expect( results[ 1 ].name ).toBe( 'abcd' );
        expect( results[ 2 ].name ).toBe( 'abcd lady' );
    } );

} );