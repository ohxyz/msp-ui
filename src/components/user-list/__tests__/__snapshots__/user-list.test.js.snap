// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with defaults 1`] = `
<div
  className="user-list"
  id={undefined}
>
  <div
    className="user-list__count"
  >
    <i
      className="material-icons"
    >
      perm_identity
    </i>
    <span
      className="user-list__count__literal"
    />
  </div>
  <div
    className="user-list__users"
  >
    <div
      className="user-strip"
    >
      <div
        className="user-strip__main"
      >
        <div
          className="user-strip__delete-icon user-strip__delete-icon--inactive"
          onClick={[Function]}
        >
          <i
            className="material-icons"
          >
            delete_forever
          </i>
        </div>
        <div
          className="user-strip__expand-icon"
          onClick={[Function]}
        >
          <i
            className="material-icons"
          >
            keyboard_arrow_down
          </i>
        </div>
        <div
          className="user-strip__top-bar"
        >
          <span
            className="user-strip__full-name"
          >
            Full Name
          </span>
          <span
            className="user-strip__top-org"
          >
            Top Orgnisation Name
          </span>
        </div>
        <div
          className="user-strip__middle-bar"
        >
          <div
            className="user-strip__org"
          >
            Orgnisation Name
          </div>
        </div>
      </div>
      <div
        className="user-strip__access-levels"
      >
        User Access Levels
      </div>
      <div
        className="user-strip__delete-user"
      >
        <div
          className="user-strip__delete-user__confirmation"
        >
          Are you sure want to delete this user?
        </div>
        <div
          className="user-strip__delete-user__buttons"
        >
          <button
            className="user-strip__delete-user__buttons__yes"
            onClick={[Function]}
          >
            Yes
          </button>
          <button
            className="user-strip__delete-user__buttons__no"
            onClick={[Function]}
          >
            No
          </button>
        </div>
      </div>
    </div>
    <div
      className="user-strip"
    >
      <div
        className="user-strip__main"
      >
        <div
          className="user-strip__delete-icon user-strip__delete-icon--inactive"
          onClick={[Function]}
        >
          <i
            className="material-icons"
          >
            delete_forever
          </i>
        </div>
        <div
          className="user-strip__expand-icon"
          onClick={[Function]}
        >
          <i
            className="material-icons"
          >
            keyboard_arrow_down
          </i>
        </div>
        <div
          className="user-strip__top-bar"
        >
          <span
            className="user-strip__full-name"
          >
            Full Name 2
          </span>
          <span
            className="user-strip__top-org"
          >
            Top Orgnisation Name 2
          </span>
        </div>
        <div
          className="user-strip__middle-bar"
        >
          <div
            className="user-strip__org"
          >
            Orgnisation Name 2
          </div>
        </div>
      </div>
      <div
        className="user-strip__access-levels"
      >
        User Access Levels
      </div>
      <div
        className="user-strip__delete-user"
      >
        <div
          className="user-strip__delete-user__confirmation"
        >
          Are you sure want to delete this user?
        </div>
        <div
          className="user-strip__delete-user__buttons"
        >
          <button
            className="user-strip__delete-user__buttons__yes"
            onClick={[Function]}
          >
            Yes
          </button>
          <button
            className="user-strip__delete-user__buttons__no"
            onClick={[Function]}
          >
            No
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
